ext-src/mem-reg.patch

```diff
@@ -0,0 +1,89 @@
++ncclResult_t ncclCommDeregister(ncclComm_t comm, void* handle);
++ncclResult_t ncclBuffIsRegistered(ncclComm_t comm, const void* buff, size_t count, bool* registered);
+ /*
+  * Send
+  *
+diff --git a/apps/nccl/src/nccl.cu b/apps/nccl/src/nccl.cu
+index a697be2..d8497e7 100644
+--- a/apps/nccl/src/nccl.cu
++++ b/apps/nccl/src/nccl.cu
+@@ -577,6 +577,67 @@ NCCL_API ncclResult_t ncclAllGather(const void* sendbuff, void* recvbuff, size_t
+   return ncclSuccess;
+ }
+ 
++NCCL_API ncclResult_t ncclCommRegister(ncclComm_t comm, void* buff, size_t size, void** handle) {
++  size_t buffBytes = size;
++  CUdeviceptr buffBasePtr;
++  MSCCLPP_CUTHROW(cuMemGetAddressRange(&buffBasePtr, &buffBytes, (CUdeviceptr)buff));
++
++  int rank = comm->comm->bootstrap()->getRank();
++  channelKey buffKey{(void*)buffBasePtr, buffBytes};
++
++  std::vector<mscclpp::RegisteredMemory> remoteMemories;
++
++  // Creating the channels
++  auto buffIt = comm->channelScratchInfos.find(buffKey);
++  if (buffIt == comm->channelScratchInfos.end()) {
++     std::vector<mscclpp::SmChannel> channels =
++          setupSmChannels(comm, comm->remoteScratchRegMemories, const_cast<void*>((void*)buffBasePtr));
++     ChannelInfo channelInfo{channels, channels, setupSmChannelDeviceHandles(channels), setupSmChannelDeviceHandles(channels)};
++     buffIt = comm->channelScratchInfos.emplace(buffKey, channelInfo).first;
++  }
++  auto sendIt = comm->channelInInfos.find(buffKey);
++  if (sendIt == comm->channelInInfos.end()) {
++      std::vector<mscclpp::SmChannel> channels =
++          setupSmChannels(comm, comm->remoteScratchRegMemories, const_cast<void*>((void*)buffBasePtr));
++
++      remoteMemories =
++          setupRemoteMemories(comm->comm, rank, (void*)buffBasePtr, buffBytes, mscclpp::Transport::CudaIpc);
++      std::vector<mscclpp::SmChannel> channels1 =
++          setupSmChannels(comm, remoteMemories, const_cast<void*>((void*)buffBasePtr));
++
++      ChannelInfo channelInfo{channels, channels1, setupSmChannelDeviceHandles(channels), setupSmChannelDeviceHandles(channels1)};
++      sendIt = comm->channelInInfos.emplace(buffKey, channelInfo).first;
++  }
++  auto recvIt = comm->channelOutInfos.find(buffKey);
++    if (recvIt == comm->channelOutInfos.end()) {
++      remoteMemories =
++          setupRemoteMemories(comm->comm, rank, (void*)buffBasePtr, buffBytes, mscclpp::Transport::CudaIpc);
++      std::vector<mscclpp::SmChannel> outChannels =
++          setupSmChannels(comm, remoteMemories, const_cast<void*>((void*)buffBasePtr));
++      ChannelInfo channelInfo{outChannels, outChannels, setupSmChannelDeviceHandles(outChannels), setupSmChannelDeviceHandles(outChannels)};
++      recvIt = comm->channelOutInfos.emplace(buffKey, channelInfo).first;
++  }
++  *handle = (void*) buffBasePtr;
++
++  return ncclSuccess;
++}
++
++NCCL_API ncclResult_t ncclCommDeregister(ncclComm_t comm, void* handle) {
++        if (comm && handle) {
++                handle = nullptr;
++        }
```

Comment:
The implementation nullifies the `handle` inside the function, but since it's passed as void*, the change doesn't reflect in the caller's context. This can lead to dangling pointers. I suggest passing handle as a void** to resolve this issue, if API change is possible

```suggestion
+NCCL_API ncclResult_t ncclCommDeregister(ncclComm_t comm, void** handle) {
+        if (comm && handle && *handle) {
+                *handle = nullptr;
+        }
```

Reply:
This is a NCCL API and cannot be changed. I am still working on some of these fixes.